src/System.Text.Json/src/System/Text/Json/Writer/Utf8JsonWriter.cs

```diff
@@ -1034,29 +1044,17 @@ private void Grow(int requiredSize)
 
             int sizeHint = Math.Max(DefaultGrowthSize, requiredSize);
 
-            Debug.Assert(BytesPending != 0);
+            Debug.Assert(_currentIndex != 0);
 
             _memory = default;
 
             if (_stream != null)
             {
                 Debug.Assert(_arrayBufferWriter != null);
 
-                _arrayBufferWriter.Advance(BytesPending);
-                BytesPending = 0;
-
-#if BUILDING_INBOX_LIBRARY
-                _stream.Write(_arrayBufferWriter.WrittenSpan);
```

Comment:
Is it feasible for us to special case streams that do not do I/O while writing (like MemoryStream) and keep such writes for those or is it best to keep the semantics of the `Utf8JsonWriter` consistent across all streams?

Reply:
I think the consistency is better.  (And if we did want to special-case, I assume we'd do more than is mentioned here: with a concrete MemoryStream, there'd be little reason to go through the array buffer writer.)